src/Zigzag_persistence/include/gudhi/filtered_zigzag_persistence.h

```diff
@@ -0,0 +1,470 @@
+  using face_key = typename Options::face_key;                  /**< Face ID type from external inputs. */
+  using filtration_value = typename Options::filtration_value;  /**< Type for filtration values. */
+  using dimension_type = typename Options::dimension_type;      /**< Type for dimension values. */
+
+  /**
+   * @brief Persistence index interval type.
+   */
+  using Index_interval = Gudhi::persistence_matrix::Persistence_interval<dimension_type,internal_key>;
+  /**
+   * @brief Persistence filtration interval type.
+   */
+  using Filtration_value_interval = Gudhi::persistence_matrix::Persistence_interval<dimension_type,filtration_value>;
+
+  /**
+   * @brief Constructor.
+   * @details After construction of the class, the zigzag filtration should be given in a streaming like way, i.e.,
+   * call @ref insert_face, @ref remove_face or @ref apply_identity for each step of the filtration in order of
+   * the filtration. To retrieve the current persistence diagram at any moment of the filtration,
+   * use @ref get_persistence_diagram or @ref get_index_persistence_diagram.
+   *
+   * @param minNumberOfFaces Minimum number of faces that will be in a complex at some point in the filtration.
+   * If the maximal number of faces is known in advance, the memory allocation can be better optimized.
+   * Default value: 0.
+   * @param ignoreCyclesAboveDim Ignores cycles in dimension larger or equal in the final diagram.
+   * If -1, no cycles are ignored. Default value: -1.
+   */
+  Filtered_zigzag_persistence_with_storage(unsigned int minNumberOfFaces = 0, int ignoreCyclesAboveDim = -1)
+      : dimMax_(ignoreCyclesAboveDim),
+        persistenceDiagram_(),
+        numArrow_(-1),
+        previousFiltrationValue_(std::numeric_limits<filtration_value>::infinity()),
+        pers_(
+            [&](dimension_type dim, internal_key birth, internal_key death) {
+              if (dimMax_ == -1 || (dimMax_ != -1 && dim < dimMax_)) {  // don't record intervals over max dim
+                persistenceDiagram_.emplace_back(birth, death, dim);
+              }
+            },
+            minNumberOfFaces) {}
+
+  /**
+   * @brief Updates the zigzag persistence diagram after the insertion of the given face.
+   *
+   * @tparam BoundaryRange Range type needing size, begin and end members.
+   * @param faceID ID representing the inserted face.
+   * @param boundary Boundary of the inserted face. The range should be composed of the IDs of all faces contained in
+   * the boundary (i.e. with non-zero coefficients), using the ID specified as `faceID` when the corresponding face
+   * was previously inserted (recall that the faces should be inserted in order of filtration).
+   * @param dimension Dimension of the inserted face.
+   * @param filtrationValue Filtration value associated to the face.
+   * Assumed to be larger or equal to previously used filtration values.
```

Comment:
In some sense, I don't think you even need filtration values to be comparable, we could use strings to name the arrows and I think it would still work.
Getting the intervals backwards if I look at the zigzag backwards looks fine to me.
We don't really have to change anything, let's just **avoid** adding an assertion that enforces this "larger or equal".

Reply:
For `Filtered_zigzag_persistence_with_storage`, the fact that the values are comparable and monotonously growing or shrinking, is used for the way the values are retrieved from the vector. In `Filtered_zigzag_persistence`, I could indeed completely remove any constrains. But will the barcodes still makes sense then?